### ISSUE TITLE
[BUG] 'NoneType' object has no attribute 'span'

### ISSUE THREAD

**Issue by VenisBeqiri** [pytube/pytube#1933]
Yesterday was working just fine and today for some reason ive been getting this error message 'NoneType' object has no attribute 'span'
python -v : 3.12.3
pytube -v : 15.0.0

heres the whole code if someone wants to look at it:

from tkinter import filedialog, StringVar
import customtkinter
from pytube import YouTube
from urllib.error import HTTPError

def startDownload():
    try:
        ytLink = link.get()
        if not ytLink:
            finishlabel.configure(text="No link provided", text_color="red")
            return
        ytObject = YouTube(ytLink, on_progress_callback=on_progress)
        
        video = ytObject.streams.get_highest_resolution()

        # Ask for file location before starting the download
        save_path = filedialog.asksaveasfilename(defaultextension=".mp4", filetypes=[("MP4 Files", "*.mp4")])
        if not save_path:
            # User cancelled the dialog
            return
        

        title.configure(text=ytObject.title, text_color="white")
        video.download(output_path=save_path)  # Add the output_path parameter
        finishlabel.configure(text="Downloaded!", text_color="green")
    except HTTPError as e:
         print(f"HTTP error occurred: {e}, URL: {ytLink}")
    except Exception as e:
        finishlabel.configure(text=f"Download Error: {str(e)}", text_color="red")

def on_progress(stream, chunk, bytes_remaining):
    total_size = stream.filesize
    bytes_downloaded = total_size - bytes_remaining
    percentage_of_completion = bytes_downloaded / total_size * 100
    per = str(int(percentage_of_completion))
    progressbar.set(percentage_of_completion)
    pPercentage.configure(text=f"{per}%")

customtkinter.set_appearance_mode("System")
customtkinter.set_default_color_theme("blue")

app = customtkinter.CTk()
app.geometry("720x480")
app.title("YouTube Downloader")

# Label
title = customtkinter.CTkLabel(app, text="Youtube Link")
title.pack(padx=10, pady=10)

#link input
url_var = StringVar()
link = customtkinter.CTkEntry(app, width=350, height=40, textvariable=url_var)
link.pack()

#Progress bar
pPercentage = customtkinter.CTkLabel(app, text="0%")
pPercentage.pack()

progressbar = customtkinter.CTkProgressBar(app, width=400)
progressbar.set(0)
progressbar.pack(padx=10, pady=10)

# Finished downloading
finishlabel = customtkinter.CTkLabel(app, text="")
finishlabel.pack()

# Download button
download = customtkinter.CTkButton(app, text="Download", command=startDownload)
download.pack(padx=10, pady=10)

app.mainloop()

![Screenshot 2024-05-22 232539](https://github.com/pytube/pytube/assets/129770545/d193bff5-e513-4ac7-83c7-da55c4e41524)

**Comment by TyrelCB** [comment]
Same here, youtube may have updated the DOM

```python
from pytube import YouTube
YouTube('https://youtu.be/2lAe1cqCOXo').streams.first().download()
````

AttributeError                            Traceback (most recent call last)
Cell In[10], [line 2](vscode-notebook-cell:?execution_count=10&line=2)
      [1](vscode-notebook-cell:?execution_count=10&line=1) from pytube import YouTube
----> [2](vscode-notebook-cell:?execution_count=10&line=2) YouTube('https://youtu.be/2lAe1cqCOXo').streams.first().download()

File ~/.local/lib/python3.12/site-packages/pytube/__main__.py:296, in YouTube.streams(self)
    [291](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:291) """Interface to query both adaptive (DASH) and progressive streams.
    [292](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:292) 
    [293](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:293) :rtype: :class:`StreamQuery <StreamQuery>`.
    [294](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:294) """
    [295](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:295) self.check_availability()
--> [296](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:296) return StreamQuery(self.fmt_streams)

File ~/.local/lib/python3.12/site-packages/pytube/__main__.py:181, in YouTube.fmt_streams(self)
    [178](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:178) # If the cached js doesn't work, try fetching a new js file
    [179](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:179) # https://github.com/pytube/pytube/issues/1054
    [180](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:180) try:
--> [181](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:181)     extract.apply_signature(stream_manifest, self.vid_info, self.js)
    [182](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:182) except exceptions.ExtractError:
    [183](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:183)     # To force an update to the js file, we clear the cache and retry
    [184](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/__main__.py:184)     self._js = None

File ~/.local/lib/python3.12/site-packages/pytube/extract.py:409, in apply_signature(stream_manifest, vid_info, js)
    [400](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/extract.py:400) def apply_signature(stream_manifest: Dict, vid_info: Dict, js: str) -> None:
...
--> [411](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/cipher.py:411) transform_plan_raw = find_object_from_startpoint(raw_code, match.span()[1] - 1)
    [413](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/cipher.py:413) # Steps are either c[x](c[y]) or c[x](c[y],c[z])
    [414](https://file+.vscode-resource.vscode-cdn.net/home/tyrel/python/music_helper/~/.local/lib/python3.12/site-packages/pytube/cipher.py:414) step_start = r"c\[(\d+)\]\(c\[(\d+)\](,c(\[(\d+)\]))?\)"

AttributeError: 'NoneType' object has no attribute 'span'

**Comment by TyrelCB** [comment]
Fix: Change line 411
from
`transform_plan_raw = find_object_from_startpoint(raw_code, match.span()[1] - 1)`
to
`transform_plan_raw = js`

**Comment by Choi-YoungHyun** [comment]
me too

yesterday pytube doing well 
but, pytube error make suddenly 


here is my code

```
def convert_youTube(youtubeURL):
    video = YouTube(youtubeURL)
    video.streams.filter(only_audio=True).first().download(output_path='.', filename= MP3_FILE)

URL =  input("Please enter Youtube URL")
 convert_youTube(URL)
```

eror

```
  File "C:\Users\choiyounghyun\AppData\Local\Programs\Python\Python39\lib\site-packages\pytube\cipher.py", line 411, in get_throttling_plan
    transform_plan_raw = find_object_from_startpoint(raw_code, match.span()[1] - 1)
AttributeError: 'NoneType' object has no attribute 'span'
```

**Comment by NeonLightning** [comment]
> me too
> 
> yesterday pytube doing well but, pytube error make suddenly
> 
> here is my code
> 
> ```
> def convert_youTube(youtubeURL):
>     video = YouTube(youtubeURL)
>     video.streams.filter(only_audio=True).first().download(output_path='.', filename= MP3_FILE)
> 
> URL =  input("Please enter Youtube URL")
>  convert_youTube(URL)
> ```
> 
> eror
> 
> ```
>   File "C:\Users\choiyounghyun\AppData\Local\Programs\Python\Python39\lib\site-packages\pytube\cipher.py", line 411, in get_throttling_plan
>     transform_plan_raw = find_object_from_startpoint(raw_code, match.span()[1] - 1)
> AttributeError: 'NoneType' object has no attribute 'span'
> ```

that line 411 in cipher.py fix works.

**Comment by Choi-YoungHyun** [comment]
> > me too
> > yesterday pytube doing well but, pytube error make suddenly
> > here is my code
> > ```
> > def convert_youTube(youtubeURL):
> >     video = YouTube(youtubeURL)
> >     video.streams.filter(only_audio=True).first().download(output_path='.', filename= MP3_FILE)
> > 
> > URL =  input("Please enter Youtube URL")
> >  convert_youTube(URL)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > eror
> > ```
> >   File "C:\Users\choiyounghyun\AppData\Local\Programs\Python\Python39\lib\site-packages\pytube\cipher.py", line 411, in get_throttling_plan
> >     transform_plan_raw = find_object_from_startpoint(raw_code, match.span()[1] - 1)
> > AttributeError: 'NoneType' object has no attribute 'span'
> > ```
> 
> that line 411 in cipher.py fix works.


Yep !
I fix line 411 in ciper.py 
about https://stackoverflow.com/questions/75765213/pytube-attributeerror-nonetype-object-has-no-attribute-span-cipher-py 
Thx

**Comment by anurag12-webster** [comment]
why this fix aint merged into master yet?

**Comment by ghost** [comment]
> Fix: Change line 411 from `transform_plan_raw = find_object_from_startpoint(raw_code, match.span()[1] - 1)` to `transform_plan_raw = js`

Fixes! THX!

**Comment by NeonLightning** [comment]
> why this fix aint merged into master yet?

because things don't happen right away?

**Comment by Andy339** [comment]
Can confirm, line 411 in cipher.py fix is working on Termux.

**Comment by s0me1code** [comment]
> Fix: Change line 411 from `transform_plan_raw = find_object_from_startpoint(raw_code, match.span()[1] - 1)` to `transform_plan_raw = js`

Thank! worked
